File: PlaySpot/Network/RestRemoteDataSource.swift
// Network/RestRemoteDataSource.swift — 신규 /api/v1/** REST 백엔드
// Phase 3-4 에서 메서드별로 실제 호출 구현. 현재는 fetch/플레이 기록만 구현.
import Foundation
import os

struct RestRemoteDataSource: MissionDataSource {
    private let client = RestAPIClient.shared
    private static let log = Logger(subsystem: "com.ejola.playspot", category: "RestDS")

    // 신규 API 가 "yyyy-MM-dd HH:mm:ss" (KST) 를 권장 (api_client.md §0).
    private static let kstFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "yyyy-MM-dd HH:mm:ss"
        f.locale = Locale(identifier: "en_US_POSIX")
        f.timeZone = TimeZone(identifier: "Asia/Seoul")
        return f
    }()

    // MARK: - 미션 목록

    func fetchMissionList(cursor: Int, lang: String) async throws -> [Mission] {
        do {
            return try await client.get("/api/v1/missions", query: ["page": "\(cursor)"])
        } catch {
            Self.log.error("fetchMissionList: \(error.localizedDescription, privacy: .public)")
            return []
        }
    }

    func fetchPublishedMissions(cursor: Int, lang: String, latitude: Double, longitude: Double) async throws -> [Mission] {
        do {
            return try await client.get("/api/v1/missions/nearby", query: [
                "page": "\(cursor)",
                "latitude": "\(latitude)",
                "longitude": "\(longitude)"
            ])
        } catch {
            Self.log.error("fetchPublishedMissions: \(error.localizedDescription, privacy: .public)")
            return []
        }
    }

    func fetchTutorialMissions(region: String) async throws -> [Mission] {
        do {
            return try await client.get("/api/v1/missions/tutorial", query: ["lang": region])
        } catch {
            Self.log.error("fetchTutorialMissions: \(error.localizedDescription, privacy: .public)")
            return []
        }
    }

    func fetchMyDesigned(userID: String) async throws -> [Mission] {
        do {
            return try await client.get("/api/v1/users/\(urlEncode(userID))/missions/designed")
        } catch {
            Self.log.error("fetchMyDesigned: \(error.localizedDescription, privacy: .public)")
            return []
        }
    }

    func fetchMyPlayed(userID: String) async throws -> [Mission] {
        do {
            return try await client.get("/api/v1/users/\(urlEncode(userID))/missions/played")
        } catch {
            Self.log.error("fetchMyPlayed: \(error.localizedDescription, privacy: .public)")
            return []
        }
    }

    func fetchCurrentGames(userID: String) async throws -> [Mission] {
        do {
            return try await client.get("/api/v1/users/\(urlEncode(userID))/missions/playing")
        } catch {
            Self.log.error("fetchCurrentGames: \(error.localizedDescription, privacy: .public)")
            return []
        }
    }

    // MARK: - 미션 상세 / 리뷰

    func fetchMissionDetail(missionID: String) async throws -> (Mission, [MissionItem], [ItemQuiz]) {
        let res: MissionDetailRes = try await client.get("/api/v1/missions/\(urlEncode(missionID))")
        // 신규 API quiz 항목은 ItemID 포함 — 정상 그룹핑.
        // 한 Quiz/Quiz20 아이템에 여러 variant(Seq=1,2,...) 또는 여러 Quiz 아이템 모두 지원.
        let quizzes = res.quizzes.map { q in
            ItemQuiz(missionID: missionID, itemID: q.ItemID, seq: q.Seq,
                     quiz: q.Quiz, answer: q.Answer, probability: q.Probability ?? 100)
        }
        return (res.mission, res.items, quizzes)
    }

    func fetchReplies(missionID: String) async throws -> [MissionReply] {
        do {
            let rows: [ReplyRes] = try await client.get("/api/v1/missions/\(urlEncode(missionID))/replies")
            return rows.compactMap { r in
                guard let text = r.MReply, !text.isEmpty else { return nil }
                let date = (r.WriteDate.flatMap { Self.parseReplyDate($0) })
                return MissionReply(text: text, score: r.Score, nickname: r.Nickname, writeDate: date)
            }
        } catch {
            Self.log.error("fetchReplies: \(error.localizedDescription, privacy: .public)")
            return []
        }
    }

    /// 서버 `WriteDate` 포맷 — 현재 `"2026-05-18T09:55:40.000+00:00"` (ISO 8601 + 밀리초 + tz).
    /// `"yyyy-MM-dd HH:mm:ss"` 도 대비해 fallback 체인 구성.
    private static func parseReplyDate(_ s: String) -> Date? {
        // 1) ISO 8601 + fractional seconds + tz
        let iso = ISO8601DateFormatter()
        iso.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let d = iso.date(from: s) { return d }
        // 2) ISO 8601 without fractional
        iso.formatOptions = [.withInternetDateTime]
        if let d = iso.date(from: s) { return d }
        // 3) "yyyy-MM-dd HH:mm:ss" KST
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US_POSIX")
        f.timeZone = TimeZone(identifier: "Asia/Seoul")
        f.dateFormat = "yyyy-MM-dd HH:mm:ss"
        return f.date(from: s)
    }

    func submitReview(missionID: String, userID: String, score: Float, reply: String) async throws -> Bool {
        let body = ReplyReq(userId: userID, score: score, reply: reply)
        do {
            try await client.send(.POST, "/api/v1/missions/\(urlEncode(missionID))/replies", body: body)
            return true
        } catch { return false }
    }

    // MARK: - 랭킹

    func fetchRanking(missionID: String) async throws -> [RankingEntry] {
        do {
            let r: RankingRes = try await client.get("/api/v1/missions/\(urlEncode(missionID))/ranking")
            var entries: [RankingEntry] = []
            let pairs = [
                (1, r.ShortUser1, r.ShortRecord1),
                (2, r.ShortUser2, r.ShortRecord2),
                (3, r.ShortUser3, r.ShortRecord3),
            ]
            for (idx, user, record) in pairs {
                guard let u = user, !u.isEmpty else { continue }
                entries.append(RankingEntry(id: idx, userName: u, record: record ?? ""))
            }
            return entries
        } catch {
            Self.log.error("fetchRanking: \(error.localizedDescription, privacy: .public)")
            return []
        }
    }

    // MARK: - 인증

    func login(email: String, password: String) async throws -> Bool {
        // password 는 평문 — 서버가 해싱 책임. HTTPS 가 전송 보호.
        do {
            let res: LoginRes = try await client.send(.POST, "/api/v1/auth/login",
                                                      body: LoginReq(userId: email, password: password))
            await AuthSession.shared.setToken(res.token)
            await AuthSession.shared.saveCredentials(userID: email, password: password)
            return true
        } catch {
            Self.log.error("login: \(error.localizedDescription, privacy: .public)")
            return false
        }
    }

    func register(email: String, password: String) async throws -> Bool {
        do {
            try await client.send(.POST, "/api/v1/auth/register",
                                  body: RegisterReq(userId: email, password: password))
            return true
        } catch let APIError.server(code, _, _, _) where code == "DUPLICATE_DATA" {
            // 이미 가입된 사용자 — register 실패지만 login 으로 진행 가능.
            return true
        } catch {
            Self.log.error("register: \(error.localizedDescription, privacy: .public)")
            return false
        }
    }

    // MARK: - 빌더

    /// @deprecated — 3-string 페이로드는 legacy 호환용. 신규 코드는 createMission 사용.
    func uploadMission(missionJSON: String, itemsJSON: String, quizzesJSON: String) async throws -> Bool {
        Self.log.warning("uploadMission: legacy 3-string payload not supported on REST backend — use createMission(_:) instead")
        return false
    }

    /// `POST /api/v1/missions` (plan_designer.md §5.2).
    /// 서버 미준비 시 404/501 발생 → 호출자가 backend 토글(.legacy) 로 폴백 권장.
    func createMission(_ req: BuilderMissionReq) async throws -> String {
        do {
            let res: BuilderMissionCreatedRes = try await client.send(.POST, "/api/v1/missions", body: req)
            return res.missionId
        } catch {
            Self.log.error("createMission: \(error.localizedDescription, privacy: .public)")
            throw error
        }
    }

    /// `PATCH /api/v1/missions/{id}` — 전체 교체.
    /// 실패 시 `APIError` rethrow — 호출자는 `error.isNotFound` / `isForbidden` / `isValidationError` 로 분기.
    func updateMission(missionID: String, _ req: BuilderMissionReq) async throws -> Bool {
        do {
            try await client.send(.PATCH, "/api/v1/missions/\(urlEncode(missionID))", body: req)
            return true
        } catch {
            Self.log.error("updateMission: \(error.localizedDescription, privacy: .public)")
            throw error
        }
    }

    /// `DELETE /api/v1/missions/{id}`.
    /// 실패 시 `APIError` rethrow — `isNotFound` (이미 삭제됨) / `isForbidden` / `isNotDeletable` (Status=2) 분기.
    func deleteMission(missionID: String) async throws -> Bool {
        do {
            try await client.send(.DELETE, "/api/v1/missions/\(urlEncode(missionID))")
            return true
        } catch {
            Self.log.error("deleteMission: \(error.localizedDescription, privacy: .public)")
            throw error
        }
    }

    /// `POST /api/v1/badges` (multipart `file`).
    /// 응답 fileName 을 받아서 호출자가 createMission/updateMission payload 의 `BadgeImageName` 에 사용.
    /// 실패 시 nil 로 삼키지 않고 throw — 호출자가 PATCH 단계 진입 전에 차단할 수 있도록.
    func uploadBadgeImage(pngData: Data) async throws -> String? {
        do {
            let res: BadgeUploadRes = try await client.uploadFile(
                "/api/v1/badges",
                fieldName: "file",
                fileName: "badge-\(UUID().uuidString.prefix(8)).png",
                mimeType: "image/png",
                data: pngData
            )
            return res.fileName
        } catch {
            Self.log.error("uploadBadgeImage: \(error.localizedDescription, privacy: .public)")
            throw error
        }
    }

    /// `POST /api/v1/files/upload` (multipart `file`) — 일반 파일 업로드.
    /// 파라미터명 `pngData` 는 historical — 실제로는 PNG/JPEG 등 어떤 바이너리든 가능.
    /// MIME 타입은 fileName 확장자에서 추론. 실패 시 throw.
    func uploadFile(pngData: Data, fileName: String) async throws -> FileUploadRes? {
        let mime: String
        let lower = fileName.lowercased()
        if lower.hasSuffix(".jpg") || lower.hasSuffix(".jpeg") { mime = "image/jpeg" }
        else if lower.hasSuffix(".webp") { mime = "image/webp" }
        else { mime = "image/png" }   // 기본 + .png
        do {
            let res: FileUploadRes = try await client.uploadFile(
                "/api/v1/files/upload",
                fieldName: "file",
                fileName: fileName,
                mimeType: mime,
                data: pngData
            )
            return res
        } catch {
            Self.log.error("uploadFile: \(error.localizedDescription, privacy: .public)")
            throw error
        }
    }

    // MARK: - 플레이 기록

    func recordPlayStart(missionID: String, playerID: String, startTime: Date, isVirtual: Bool) async throws -> Bool {
        let body = PlayReq(playerId: playerID,
                           startTime: Self.kstFormatter.string(from: startTime),
                           endTime: nil,
                           isVirtual: isVirtual ? 1 : 0)
        do {
            let _: PlayResultRes = try await client.send(.POST, "/api/v1/missions/\(urlEncode(missionID))/plays/start", body: body)
            return true
        } catch {
            // 200 + {result:SUCCESS} 가 아닌 204 응답 가능성도 흡수.
            do {
                try await client.send(.POST, "/api/v1/missions/\(urlEncode(missionID))/plays/start", body: body)
                return true
            } catch {
                Self.log.error("recordPlayStart: \(error.localizedDescription, privacy: .public)")
                return false
            }
        }
    }

    func recordPlayFinish(missionID: String, playerID: String, startTime: Date, endTime: Date, isVirtual: Bool) async throws -> Bool {
        let body = PlayReq(playerId: playerID,
                           startTime: Self.kstFormatter.string(from: startTime),
                           endTime: Self.kstFormatter.string(from: endTime),
                           isVirtual: isVirtual ? 1 : 0)
        do {
            try await client.send(.POST, "/api/v1/missions/\(urlEncode(missionID))/plays/finish", body: body)
            return true
        } catch {
            Self.log.error("recordPlayFinish: \(error.localizedDescription, privacy: .public)")
            return false
        }
    }

    func recordPlayFail(missionID: String, playerID: String, startTime: Date, endTime: Date, isVirtual: Bool) async throws -> Bool {
        let body = PlayReq(playerId: playerID,
                           startTime: Self.kstFormatter.string(from: startTime),
                           endTime: Self.kstFormatter.string(from: endTime),
                           isVirtual: isVirtual ? 1 : 0)
        do {
            try await client.send(.POST, "/api/v1/missions/\(urlEncode(missionID))/plays/fail", body: body)
            return true
        } catch {
            Self.log.error("recordPlayFail: \(error.localizedDescription, privacy: .public)")
            return false
        }
    }

    // MARK: - User

    func fetchUser(userID: String) async throws -> UserRes? {
        do {
            return try await client.get("/api/v1/users/\(urlEncode(userID))")
        } catch {
            Self.log.error("fetchUser: \(error.localizedDescription, privacy: .public)")
            return nil
        }
    }

    func updateUser(userID: String, patch: UserPatchReq) async throws -> Bool {
        do {
            try await client.send(.PATCH, "/api/v1/users/\(urlEncode(userID))", body: patch)
            return true
        } catch {
            Self.log.error("updateUser: \(error.localizedDescription, privacy: .public)")
            return false
        }
    }

    func changePassword(userID: String, oldPasswordMD5: String, newPasswordMD5: String) async throws -> Bool {
        let body = PasswordChangeReq(oldPassword: oldPasswordMD5, newPassword: newPasswordMD5)
        do {
            try await client.send(.PATCH, "/api/v1/users/\(urlEncode(userID))/password", body: body)
            return true
        } catch {
            Self.log.error("changePassword: \(error.localizedDescription, privacy: .public)")
            return false
        }
    }

    // MARK: - utility

    private func urlEncode(_ s: String) -> String {
        s.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? s
    }
}

File: flutter_ar_spike/lib/network/rest_remote_data_source.dart
// network/rest_remote_data_source.dart — RestRemoteDataSource.swift 이식 (/api/v1/**).
import 'dart:developer' as dev;
import 'package:flutter/foundation.dart' show kDebugMode;
import '../models/mission.dart';
import '../models/mission_item.dart';
import '../models/item_quiz.dart';
import '../models/mission_reply.dart';
import '../models/ranking_entry.dart';
import 'auth_session.dart';
import 'builder_mission_req.dart';
import 'mission_data_source.dart';
import 'rest_api_client.dart';

class RestRemoteDataSource implements MissionDataSource {
  final RestApiClient _client;
  final AuthSession _auth;
  RestRemoteDataSource(this._client, this._auth);

  List<Mission> _missions(dynamic data) =>
      (data as List<dynamic>? ?? [])
          .map((e) => Mission.fromJson(e as Map<String, dynamic>))
          .toList();

  void _log(String m) => dev.log(m, name: 'RestDS');

  // 읽기 — 실패 시 [] (Swift 동일, 게임 흐름 차단 방지).
  @override
  Future<List<Mission>> fetchMissionList({int cursor = 0, String lang = ''}) async {
    try {
      return _missions(await _client.get('/api/v1/missions', query: {'page': '$cursor'}));
    } catch (e) {
      _log('fetchMissionList: $e');
      return [];
    }
  }

  @override
  Future<List<Mission>> fetchPublishedMissions(
      {int cursor = 0, String lang = '', required double latitude, required double longitude}) async {
    try {
      return _missions(await _client.get('/api/v1/missions/nearby',
          query: {'page': '$cursor', 'latitude': '$latitude', 'longitude': '$longitude'}));
    } catch (e) {
      _log('fetchPublishedMissions: $e');
      return [];
    }
  }

  @override
  Future<List<Mission>> fetchTutorialMissions(String region) async {
    try {
      return _missions(await _client.get('/api/v1/missions/tutorial', query: {'lang': region}));
    } catch (e) {
      _log('fetchTutorialMissions: $e');
      return [];
    }
  }

  @override
  Future<List<Mission>> fetchMyDesigned(String userID) async {
    try {
      return _missions(await _client.get('/api/v1/users/${Uri.encodeComponent(userID)}/missions/designed'));
    } catch (e) {
      _log('fetchMyDesigned: $e');
      return [];
    }
  }

  @override
  Future<List<Mission>> fetchMyPlayed(String userID) async {
    try {
      return _missions(await _client.get('/api/v1/users/${Uri.encodeComponent(userID)}/missions/played'));
    } catch (e) {
      _log('fetchMyPlayed: $e');
      return [];
    }
  }

  @override
  Future<List<Mission>> fetchCurrentGames(String userID) async {
    try {
      return _missions(await _client.get('/api/v1/users/${Uri.encodeComponent(userID)}/missions/playing'));
    } catch (e) {
      _log('fetchCurrentGames: $e');
      return [];
    }
  }

  @override
  Future<MissionDetail> fetchMissionDetail(String missionID) async {
    final data = await _client.get('/api/v1/missions/${Uri.encodeComponent(missionID)}') as Map<String, dynamic>;
    final mission = Mission.fromJson(data['mission'] as Map<String, dynamic>);
    final items = (data['items'] as List<dynamic>? ?? [])
        .map((e) => MissionItem.fromJson(e as Map<String, dynamic>))
        .toList();
    final quizzes = (data['quizzes'] as List<dynamic>? ?? []).map((e) {
      final q = e as Map<String, dynamic>;
      return ItemQuiz(
        missionID: missionID,
        itemID: (q['ItemID'] as num).toInt(),
        seq: (q['Seq'] as num).toInt(),
        quiz: q['Quiz'] as String? ?? '',
        answer: q['Answer'] as String? ?? '',
        probability: (q['Probability'] as num?)?.toInt() ?? 100,
      );
    }).toList();
    return (mission, items, quizzes);
  }

  @override
  Future<List<MissionReply>> fetchReplies(String missionID) async {
    try {
      final rows = await _client.get('/api/v1/missions/${Uri.encodeComponent(missionID)}/replies') as List<dynamic>? ?? [];
      return rows
          .map((e) => MissionReply.fromJson(e as Map<String, dynamic>))
          .whereType<MissionReply>()
          .toList();
    } catch (e) {
      _log('fetchReplies: $e');
      return [];
    }
  }

  @override
  Future<List<RankingEntry>> fetchRanking(String missionID) async {
    try {
      final r = await _client.get('/api/v1/missions/${Uri.encodeComponent(missionID)}/ranking') as Map<String, dynamic>;
      final entries = <RankingEntry>[];
      final pairs = [
        (1, r['ShortUser1'] as String?, r['ShortRecord1'] as String?),
        (2, r['ShortUser2'] as String?, r['ShortRecord2'] as String?),
        (3, r['ShortUser3'] as String?, r['ShortRecord3'] as String?),
      ];
      for (final (idx, user, record) in pairs) {
        if (user == null || user.isEmpty) continue;
        entries.add(RankingEntry(id: idx, userName: user, record: record ?? ''));
      }
      return entries;
    } catch (e) {
      _log('fetchRanking: $e');
      return [];
    }
  }

  @override
  Future<bool> submitReview(
      {required String missionID, required String userID, required double score, required String reply}) async {
    try {
      await _client.send('POST', '/api/v1/missions/${Uri.encodeComponent(missionID)}/replies',
          body: {'userId': userID, 'score': score, 'reply': reply});
      return true;
    } catch (e) {
      _log('submitReview: $e');
      return false;
    }
  }

  // 인증
  /// 디버그 빌드 한정 테스트 계정 (서버 우회).
  static const _devEmail = 'test@gmail.com';
  static const _devPassword = '1234';
  bool _isDevAccount(String e, String p) =>
      kDebugMode && e == _devEmail && p == _devPassword;

  @override
  Future<bool> login(String email, String password) async {
    // 디버그 테스트 계정 — 서버 호출 없이 즉시 성공.
    if (_isDevAccount(email, password)) {
      await _auth.setToken('dev_test_token');
      await _auth.saveCredentials(email, password);
      _auth.userId = email;
      _log('login: dev bypass for $email');
      return true;
    }
    try {
      final res = await _client.send('POST', '/api/v1/auth/login',
          body: {'userId': email, 'password': password}) as Map<String, dynamic>;
      final token = res['token'] as String?;
      if (token == null) return false;
      await _auth.setToken(token);
      await _auth.saveCredentials(email, password);
      _auth.userId = email; // setter 가 notifyListeners → Settings 등 자동 갱신
      return true;
    } catch (e) {
      _log('login: $e');
      return false;
    }
  }

  @override
  Future<bool> register(String email, String password) async {
    // 디버그 테스트 계정 — 서버 호출 없이 즉시 가입 성공 (이후 login 호출도 통과).
    if (_isDevAccount(email, password)) {
      _log('register: dev bypass for $email');
      return true;
    }
    try {
      await _client.send('POST', '/api/v1/auth/register',
          body: {'userId': email, 'password': password});
      return true;
    } catch (e) {
      // DUPLICATE_DATA(이미 가입) 는 login 진행 가능 → true 로 흡수.
      if (e.toString().contains('DUPLICATE_DATA')) return true;
      _log('register: $e');
      return false;
    }
  }

  // 빌더 CUD — 실패 시 rethrow (호출부에서 분기).
  @override
  Future<String> createMission(BuilderMissionReq req) async {
    final res = await _client.send('POST', '/api/v1/missions', body: req.toJson()) as Map<String, dynamic>;
    return res['missionId'] as String;
  }

  @override
  Future<bool> updateMission(String missionID, BuilderMissionReq req) async {
    await _client.send('PATCH', '/api/v1/missions/${Uri.encodeComponent(missionID)}', body: req.toJson());
    return true;
  }

  @override
  Future<bool> deleteMission(String missionID) async {
    await _client.send('DELETE', '/api/v1/missions/${Uri.encodeComponent(missionID)}');
    return true;
  }

  /// SwiftUI RestRemoteDataSource.uploadFile 1:1 — POST /api/v1/files/upload (multipart `file`).
  /// 응답 `fileUrl` (S3 전체 https URL) 을 그대로 반환. mission.badgeImageName 에 저장하면 됨.
  /// `BadgeImageName: "https://playspot-badge-dev.s3.amazonaws.com/file/..."`
  @override
  Future<String?> uploadFile(List<int> bytes, String fileName) async {
    final lower = fileName.toLowerCase();
    final mime = lower.endsWith('.jpg') || lower.endsWith('.jpeg')
        ? 'image/jpeg'
        : lower.endsWith('.webp')
            ? 'image/webp'
            : 'image/png';
    final res = await _client.uploadFile('/api/v1/files/upload',
        fieldName: 'file', fileName: fileName, mimeType: mime, bytes: bytes);
    if (res is Map) {
      final url = (res['fileUrl'] as String?) ?? (res['fileName'] as String?);
      if (url != null && url.isNotEmpty) return url;
    }
    _log('uploadFile: unexpected response: $res');
    return null;
  }

  /// 레거시 `POST /api/v1/badges` — 호환용. 신규 흐름은 uploadFile 사용.
  @override
  Future<String?> uploadBadgeImage(List<int> bytes, String fileName) async {
    final lower = fileName.toLowerCase();
    final mime = lower.endsWith('.jpg') || lower.endsWith('.jpeg')
        ? 'image/jpeg'
        : lower.endsWith('.webp')
            ? 'image/webp'
            : 'image/png';
    try {
      final res = await _client.uploadFile('/api/v1/badges',
          fieldName: 'file', fileName: fileName, mimeType: mime, bytes: bytes);
      if (res is Map) {
        final n = (res['fileName'] as String?) ??
            (res['fileUrl'] as String?) ??
            (res['url'] as String?);
        if (n != null && n.isNotEmpty) return n;
      }
      _log('uploadBadgeImage: unexpected response shape: $res');
      return null;
    } catch (e) {
      _log('uploadBadgeImage: $e');
      rethrow; // 호출자가 사용자에게 정확한 에러 메시지 전달.
    }
  }

  // 플레이 기록 — KST "yyyy-MM-dd HH:mm:ss", best-effort.
  static String _kst(DateTime d) {
    final l = d.toLocal();
    String two(int n) => n.toString().padLeft(2, '0');
    return '${l.year}-${two(l.month)}-${two(l.day)} ${two(l.hour)}:${two(l.minute)}:${two(l.second)}';
  }

  Future<bool> _recordPlay(String missionID, String path, Map<String, dynamic> body) async {
    try {
      await _client.send('POST', '/api/v1/missions/${Uri.encodeComponent(missionID)}/plays/$path', body: body);
      return true;
    } catch (e) {
      _log('recordPlay/$path: $e');
      return false;
    }
  }

  @override
  Future<bool> recordPlayStart(
      {required String missionID, required String playerID, required DateTime startTime, required bool isVirtual}) {
    return _recordPlay(missionID, 'start',
        {'playerId': playerID, 'startTime': _kst(startTime), 'isVirtual': isVirtual ? 1 : 0});
  }

  @override
  Future<bool> recordPlayFinish(
      {required String missionID, required String playerID, required DateTime startTime, required DateTime endTime, required bool isVirtual}) {
    return _recordPlay(missionID, 'finish',
        {'playerId': playerID, 'startTime': _kst(startTime), 'endTime': _kst(endTime), 'isVirtual': isVirtual ? 1 : 0});
  }

  @override
  Future<bool> recordPlayFail(
      {required String missionID, required String playerID, required DateTime startTime, required DateTime endTime, required bool isVirtual}) {
    return _recordPlay(missionID, 'fail',
        {'playerId': playerID, 'startTime': _kst(startTime), 'endTime': _kst(endTime), 'isVirtual': isVirtual ? 1 : 0});
  }
}
